File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {

        List<Integer> ans = new ArrayList<>();

        // we cal a helper fucntion where it does this thing
        preOrder(root,ans);

        return ans;


    }

    // helper funciton (IMP)
    public void preOrder(TreeNode root, List<Integer> ans){

        // if the current node is a function(recursive function ) is null then go back to its previous fucntion where this originated
        if(root == null){
            return;
        }

        // according to pre order traversal -> first print the node -> then left subtree -> then right subtree
        ans.add(root.val);
        // now go to the left subtree and so on (if that node has children -> then to its left subtree ans so on...)
        preOrder(root.left,ans);
        // after finishing the left paart now the right part starts
        preOrder(root.right,ans);

    }
}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        // code is done using recurision
        // first for every node if the node is null , then return 0
        if(root == nullptr){
            return 0;
        }

        // now first calculate the depth for left subtree
        int leftSide = maxDepth(root->left);
        // then for right subtree
        int rightSide = maxDepth(root->right);
        // after that this is the main formula:
        return max(leftSide,rightSide) + 1;
        // The reason we add +1 after finding the maximum depth of the left and right subtrees is to account for the current node itself

    }
};